### ISSUE TITLE
Cohort caching fails on RedShift

### ISSUE THREAD

**Issue by anthonysena** [OHDSI/WebAPI#1455]
Cohort generation on RedShift fails due to the new caching mechanism with the following error:

````
Details: 
 -----------------------------------------------
  error:  Value out of range for 4 bytes.
  code:      8001
  context:   Input:97887652888755345.
  query:     19121631
  location:  numeric_bound.cpp:43
  process:   query0_975_19121631 [pid=47705]
  -----------------------------------------------;
````

After tracing through the offending SQL, it appears that the `AVG` aggregate will overflow when we cast to a BIGINT. I've created a work-around that works well for reference: https://github.com/OHDSI/WebAPI/commit/d70eb8e41b1f6c891f58c7d401bf7a4e54257727. We  need to consider a more permanent fix for this.

**Comment by anthonysena** [comment]
@olga-ganina can someone on your team take a look to get your thoughts on potential fixes? As I've noted here, I've made a hack to work around it but would like to see if we can find a more permanent solution.

**Comment by wivern** [comment]
I guess the root of the issue is `HASHBYTES` function that returns 32 characters value but not the 32 bits. And then for a sql server the `CONVERT` function is used to reduce this value to int. For SqlServer convert truncates symbols from the right when result length exceeds expected (integer), but redshit's `convert` does not.
Still in research for a solution suitable for all supported by WebAPI databases.

**Comment by wivern** [comment]
@anthonysena As `convert as int` was used here just to reduce computed hash into integer size, then why not to do the same with a bitwise AND operator?
Then the line could be modified like the following:
```
CAST(CAST(CONVERT(VARBINARY, HASHBYTES('MD5',line), 1) AS BIGINT) & 2147483647 AS BIGINT)
```
Could you test this approach on RedShift?

**Comment by anthonysena** [comment]
@wivern thanks for digging into this and apologies for the delayed response. Part of the delay is that I wanted to dig into the details around this particular query to see why we're seeing different results on RedShift and so I've done a bit of that and wanted to share my findings.

Focusing on the final line from this change: https://github.com/OHDSI/WebAPI/blob/d70eb8e41b1f6c891f58c7d401bf7a4e54257727/src/main/resources/resources/generationcache/cohort/resultsChecksum.sql#L65:

````
SELECT AVG(
  CAST(CAST(CONVERT(VARBINARY, HASHBYTES('MD5',line), 1) AS NUMERIC(20,0)) AS NUMERIC(20,0))
) as checksum
````

I changed the CAST statements to use `NUMERIC(20,0)` as a work-around which seemed a bit odd. So, I ran this through SqlRender (fyi, there is an OHDSI ShinyApp for this: https://data.ohdsi.org/SqlDeveloper/), you can see the translation to RedShift is as follows:

````
SELECT AVG(
  CAST(CAST(CONVERT(VARBINARY, MD5(line), 1) AS NUMERIC(20,0)) AS NUMERIC(20,0))
) as checksum
````

What was interesting is that `CONVERT(VARBINARY, MD5(line), 1)` was not translated properly. Looking at the SqlRender rule here: https://github.com/OHDSI/SqlRender/blob/master/inst/csv/replacementPatterns.csv#L275, we can see that it is not translated because it does not match the pattern exactly. I manually edited the query a bit to see if I could make it work based on my interpretation of the replacement rule. Here is a simplistic example:

````
WITH partials AS (
  SELECT CAST('123483278327842378468273623897423948792387429361239813981279127834289365' AS VARCHAR) AS line
)
SELECT AVG(
  CAST(CAST(STRTOL(LEFT(MD5(line), 15), 16)  AS BIGINT) AS BIGINT)
) as checksum
FROM partials;
````
**NOTE**: I changed the CASTs above to use BIGINT vs NUMERIC(20,0) on RedShift and this now works.
 
I'd suggest that we revisit the changes made here: https://github.com/OHDSI/SqlRender/pull/186. I think that the translation rules for `CONVERT(VARBINARY, CONCAT('0x', @a), 1)` should be revised to `CONVERT(VARBINARY, @a, 1)`. I'm not sure we need the "0x" prefix anymore but am unsure about some of the other platforms listed in https://github.com/OHDSI/SqlRender/issues/185. Best I can tell, we're only utilizing the `HASHBYTES` via WebAPI here: https://github.com/OHDSI/WebAPI/search?q=HASHBYTES&unscoped_q=HASHBYTES. Since we're no longer requiring the "0x" prefix I think we need to revise this rule.

Here are the relevant changes from WebAPI for reference:
 https://github.com/OHDSI/WebAPI/pull/1318/commits/618065b83c1937513b977e3688dcc7a76ec3d36b

https://github.com/OHDSI/WebAPI/commit/c515fbd4c35a9164e66c8c25bb447169d1b3b310

Let me know your thoughts and if supportive, we can propose a change for this via SqlRender.

**Comment by anthonysena** [comment]
@wivern - sorry I did not respond to your earlier idea but I don't think can do a bitwise AND. After digging into this issue, I'd like to propose a different course of action from modifying the rules in SqlRender. First though, let me recap some of the discussion from #1088 that led to the current approach.

#### Design Thinking

The aim with creating the checksum operation that is causing the problem in this issue (for reference the query is [here](https://github.com/OHDSI/WebAPI/blob/master/src/main/resources/resources/generationcache/cohort/resultsChecksum.sql) was to ensure that when a given cohort design is generated on a source, we have a mechanism to verify that the contents of that generation have not changed so that we can use the cached version. 

To break down the query mentioned a bit more:

````
(... Bunch of CTEs to obtain the unique elements of a cohort generation ommitted ...)
SELECT AVG(
  CAST(CAST(
	CONVERT(VARBINARY, HASHBYTES('MD5',line), 1) 
  AS INT) AS BIGINT)
) as checksum
FROM partials;
````

The query is doing the following: for the resulting cohort generation, identify each person in the cohort, their start/end date, the inclusion rule counts, etc and each of these produces a `line` which is fed into the query above. From there the HASH of each line is calculated and then the HASH is converted to a string representation as part of: `CONVERT(VARBINARY, HASHBYTES('MD5',line), 1)`. This has the potential to produce tons of records so the results are then `CAST` first to an `INT` and then a `BIGINT` presumably to prevent overflow as part of the `AVG` operation.


#### Alternative approaches

The principle of "make sure the results have not changed" is important and one I'd like to stick to here. That said, the current approach for computing the checksum feels like it will be too complex for all of the platforms that are listed in OHDSI/SqlRender#217 and I think this issue we're facing on RedShift will appear in a different form in the other DBs.

The good news is that the WebAPI's `generation_cache.result_checksum` is a VARCHAR(255) so we have some flexibility in what we store. My current thinking is to perform a simple `COUNT_BIG` on the cohort table for the selected `design_hash` and then store that as the checksum. While this is not technically a checksum, I think that checking the row count of the cohort table should suffice to fulfill the original aims of this design and this is easily supported by all DB platforms. 

Tagging @chrisknoll @wivern for any input.

**Comment by wivern** [comment]
From my point of view, the modification of the `result_checksum` to `VARCHAR` could be a good compromise between cross-platform support and performance.
One thing I'd like to add that would be nice to make some benchmarking to know exactly how the performance has changed.

**Comment by anthonysena** [comment]
Thanks @wivern - I reviewed this with @chrisknoll as well and for now we're going to use the simplistic checksum approach I outlined above which should provide broader support across DBMS platforms as well as improved performance. The original approach developed is also very performant but seems to add complexity that will create challenges across different DB platforms. For now the simplified approach should get us to a better spot. The downside to this approach is that it does not actually use the contents of the `cohort_cache` to ensure that the contents have changed. If that becomes a hard requirement down the line, we can revisit this decision. Thanks!
